README
CONTENTS

1. DESCRIPTION
2. INSTALLATION
3. UPGRADING
4. DOCUMENTATION
5. MAILING LIST
6. CONFIGURATION
7. COMPATIBILITY
8. FEATURES
9. COPYRIGHT


--------------
1. DESCRIPTION
--------------

Convert::Binary::C is a preprocessor and parser for C type definitions.
It is highly configurable and should support arbitrarily complex data
structures. Its object-oriented interface has pack and unpack methods
that act as replacements for Perl's pack and unpack and allow to use the
C types instead of a string representation of the data structure for
conversion of binary data from and to Perl's complex data structures.

Actually, what Convert::Binary::C does is not very different from what
a C compiler does, just that it doesn't compile the source code into
an object file or executable, but only parses the code and allows Perl
to use the enumerations, structs, unions and typedefs that have been
defined within your C source for binary data conversion, similar to
Perl's pack and unpack.

Beyond that, the module offers a lot of convenience methods to retrieve
information about the C types that have been parsed.

---------------
2. INSTALLATION
---------------

Installation of the Convert::Binary::C module follows the standard
Perl Way and should not be harder than:

  perl Makefile.PL
  make
  make test
  make install

Note that you may need to become superuser to 'make install'.

If you're building the module under Windows, you may need to use a
different make program, such as 'nmake', instead of 'make'.

When running 'make test' on on slower systems please be patient,
since some of the tests are quite time consuming. The time required
for running the whole test suite depends on your Perl version, the
features you're building the module with, and of course your machine.

------------
3. UPGRADING
------------

If you are upgrading from a previous release of Convert::Binary::C,
please check the Changes file. In its current state, the module is
still subject to changes that may affect compatibility with older
releases.

----------------
4. DOCUMENTATION
----------------

To see the documentation, use the perldoc command:

  perldoc Convert::Binary::C
  perldoc Convert::Binary::C::Cached

You can also visit CPAN Search and see the documentation online as
pretty nice HTML. This is also where you will find the most recent
version of this module:

  http://search.cpan.org/~mhx/Convert-Binary-C/

Even though the documentation contains some example code, you might
want some working example scripts. You can find them in the

  examples

subdirectory after you've built the module. These scripts normally
require Convert::Binary::C to be installed on your system. If you
want to test the examples prior to installing Convert::Binary::C,
you can start the examples like this after building the module:

  perl -Mblib examples/script.pl

Otherwise just run the example scripts like any other Perl script.

---------------
5. MAILING LIST
---------------

There's also a mailing list that you can join:

  convert-binary-c@yahoogroups.com

To subscribe, simply send mail to:

  convert-binary-c-subscribe@yahoogroups.com

----------------
6. CONFIGURATION
----------------

Configuring a Convert::Binary::C object correctly can be quite
painful if you don't know every little detail about your compiler.

However, if you're lucky, you can use the 'ccconfig' tool that
comes with this package. It was written to automatically retrieve
the complete compiler configuration. This may not work always,
or retrieve the complete configuration, but it should at least
give you some point to start from.

Just run

  ccconfig -c compiler

with 'compiler' being the name of your compiler executable.

You can see the documentation for 'ccconfig' using the perldoc
command:

  perldoc ccconfig

As the tool is very experimental, any feedback on 'ccconfig'
is really appreciated!

----------------
7. COMPATIBILITY
----------------

The module should build on most of the platforms that Perl runs on.
I have tested it on:

  - Various Linux systems
  - Various BSD systems
  - HP-UX
  - Compaq/HP Tru64 Unix
  - Mac-OS X
  - Cygwin
  - Windows 98/NT/2000/XP

Also, most architectures should be supported. Tests were done on:

  - Various Intel Pentium and Itanium systems
  - Various Alpha systems
  - HP PA-RISC
  - Power-PC
  - StrongARM (the module worked fine on an IPAQ system)

The module should build with any perl from 5.005 up to the latest
development version. It will also build with perl 5.004, but then
the test suite cannot be run completely.

Multithreaded perl binaries are explicitly supported, as the module
is intended to be thread-safe.

-----------
8. FEATURES
-----------

You can enable or disable certain features at compile time by adding
options to the Makefile.PL call. However, you can safely leave them
at their default.

Available features are 'debug' to build the module with debugging
support and 'ieeefp' to explicitly enable or disable IEEE floating
point support.

The 'debug' feature depend on how your perl binary was built. If it
was built with the DEBUGGING flag, the 'debug' feature is enabled,
otherwise it is disabled by default.

The 'ieeefp' feature depends on how your machine stores floating point
values. If they are stored in IEEE format, this feature will be enabled
automatically. You should really only change the default if you know
what you're doing.

You can enable or disable features explicitly by adding the arguments

  enable-feature
  disable-feature

to the Makefile.PL call. To explicitly build the module with debugging
enabled, you would say:

  perl Makefile.PL enable-debug

This will still allow you to pass other 'standard' arguments to
Makefile.PL, like

  perl Makefile.PL enable-debug OPTIMIZE=-O3

------------
9. COPYRIGHT
------------

Copyright (c) 2002-2011 Marcus Holland-Moritz. All rights reserved.
This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

The ucpp library is (c) 1998-2002 Thomas Pornin. For license
and redistribution details refer to 'ctlib/ucpp/README'.

Portions copyright (c) 1989, 1990 James A. Roskind.

The include files located in 'tests/include/include', which are used in
some of the test scripts are (c) 1991-1999, 2000, 2001 Free Software
Foundation, Inc. They are neither required to create the binary nor
linked to the source code of this module in any other way.



Changes
----------------------------
Version 0.82 (2020-11-19)
----------------------------

* fix speed tests (the new smaller code was parsing too
  quickly for caching to have a strong impact)

* update documentation, e.g. bugtracker

----------------------------
Version 0.81 (2020-11-18)
----------------------------

* fix parallel builds

----------------------------
Version 0.80 (2020-11-17)
----------------------------

* fix 602_threads.t

----------------------------
Version 0.79_01 (2020-11-15)
----------------------------

* [WIP] fixing cpan #132904: License Inconsitencies

* update documentation

-------------------------
Version 0.79 (2020-05-19)
-------------------------

* fix cpan #121039: Fails tests when no "." in @INC; thanks to
  Kent Fredric for the report & pull request and sorry for the
  long time this bug/pr sat in the queue

* merge github PR #5: Remove use of indirect object notation in
  tests; thanks to Julien Fiegehenn for the PR and sorry for the
  long time this sat in the queue

* fix some tests that generated spurious output and didn't
  actually test anything useful

-------------------------
Version 0.78 (2016-02-14)
-------------------------

* added Clang support to ccconfig

* fix cpan #109978: improve wording/grammar in POD; thanks to
  gregor herrmann for providing a patch

* fix cpan #95716: Add support for #warn/#warning; thanks to
  Jon DuSaint for providing an initial patch

* update module metadata

-------------------------
Version 0.77 (2015-07-22)
-------------------------

* ensure consistent order of enums when generating code; this was
  a real bug and caused the configuration part of C::B::C to report
  wrong (even invalid) configurations back to the user when built
  using a recent version of Perl that randomises hash keys

* fix cpan #85264: the POD test was reporting a missing =back tag
  in one of the support modules; the ultimate fix was to ensure only
  the provided modules were POD checked

* fix cpan #83167: an artificially low refcnt for immortal variables
  introduced recently for debug Perl builds was causing the refcnt
  tests to fail for SvNULLs; thanks to Reini Urban for his patch;
  also thanks to Nicholas Clark for pointing me to the Perl commit

* fix cpan #78165: spelling errors in the documentation; thanks to
  gregor herrmann for providing a patch

* fix deprecated use of unescaped braces in regex

* fix some gcc and Perl warnings

* switch from SnapshotCM to git repo: github.com/mhx/Convert-Binary-C/

-------------------------
Version 0.76 (2011-04-11)
-------------------------

* it looks like I've found the source of the test failures reported
  in cpan #56369; this version is an attempt to fix the problem, but
  without fully understanding it -- if a windows guru can explain to
  me why setjmp() deadlocks if the jmp_buf structure was allocated
  through Perl's API (New, Newz, etc), I'd really appreciate it

-------------------------
Version 0.75 (2011-04-10)
-------------------------

* fix cpan #63381 / perl #79854: make sure warnings cause a cache
  file not to be used (thanks to Andreas König for reporting this)

* fix bug in unpack where the code would attempt to create an array
  with negative size when running out of input data

* make sure elf.PL reads/writes in binary mode (thanks to רחלי מרל
  for finding this problem)

-------------------------
Version 0.74 (2009-04-18)
-------------------------

* fix sourcification of flexible array members (triggered by cpan
  #45105; thanks to Alexander Ost for being persistent in his
  attempt to convince me that there's a bug in sourcify)

-------------------------
Version 0.73 (2009-03-16)
-------------------------

* fix META.yml generation with newer versions of ExtUtils::MakeMaker

* work around an include path handling bug present in some versions
  of Test::Harness

-------------------------
Version 0.72 (2009-03-15)
-------------------------

* fix cpan #43119: patch to allow Convert::Binary::C to build
  with GCC 4.4.x (thanks to Alex Lancaster for providing a patch)

-------------------------
Version 0.71 (2008-04-15)
-------------------------

* added new 'StdCVersion' and 'HostedC' options for configuring
  the predefined macros of the preprocessor

* fix debugging build with development version of Perl

-------------------------
Version 0.70 (2007-12-24)
-------------------------

-- released without changes --

----------------------------
Version 0.69_01 (2007-12-08)
----------------------------

* fix UV generation on platforms where NVs don't preserve
  UVs (e.g. amd64) and perl doesn't know about UVs (<5.6.0)

* make the test suite more noisy when something goes wrong

-------------------------
Version 0.69 (2007-12-06)
-------------------------

* fix: unpack would return garbage when an unpack hook
  causes the perl stack to be reallocated

* make ccconfig handle blanks in paths correctly

* slightly improve ELF parser example

* fix debug format strings

-------------------------
Version 0.68 (2007-06-25)
-------------------------

* allow #pragma pack(0) to reset to default alignment,
  just like #pragma pack

* add checks for invalid #pragma pack() arguments

-------------------------
Version 0.67 (2006-12-29)
-------------------------

* bleeding edge perl versions don't have SVt_PVBM anymore,
  which broke compilation of cbc/util.c

-------------------------
Version 0.66 (2006-11-02)
-------------------------

* fix a bug in the parser that accidentially allowed
  qualified enum specifiers as unnamed compound members

* update test suite for better coverage of recent
  compiler features and bitfields

-------------------------
Version 0.65 (2006-08-27)
-------------------------

* fix compilation issues with perls configured with
  threads and debugging; these issues were caused by
  a buggy implementation of the assert() macro in the
  5.8 series of perl

* fix compilation issues with HP compiler on itanium
  systems


cbc/cbc.h
/*******************************************************************************
*
* HEADER: cbc.h
*
********************************************************************************
*
* DESCRIPTION: C::B::C common defines
*
********************************************************************************
*
* $Project: /Convert-Binary-C $
* $Author: mhx $
* $Date: 2011/04/10 11:32:13 +0100 $
* $Revision: 29 $
* $Source: /cbc/cbc.h $
*
********************************************************************************
*
* Copyright (c) 2002-2011 Marcus Holland-Moritz. All rights reserved.
* This program is free software; you can redistribute it and/or modify
* it under the same terms as Perl itself.
*
*******************************************************************************/

#ifndef _CBC_CBC_H
#define _CBC_CBC_H

/*===== GLOBAL INCLUDES ======================================================*/


/*===== LOCAL INCLUDES =======================================================*/

#include "ctlib/arch.h"
#include "ctlib/ctdebug.h"
#include "ctlib/ctparse.h"
#include "ctlib/cttype.h"

#include "cbc/basic.h"


/*===== DEFINES ==============================================================*/

#define XSCLASS "Convert::Binary::C"

#define ARGTYPE_PACKAGE "Convert::Binary::C::ARGTYPE"

/*-------------------------------------*/
/* some quick paranoid checks first... */
/*-------------------------------------*/

#if (defined I8SIZE && I8SIZE != 1) || \
    (defined U8SIZE && U8SIZE != 1)
#error "Your I8/U8 doesn't seem to have 8 bits..."
#endif

#if (defined I16SIZE && I16SIZE != 2) || \
    (defined U16SIZE && U16SIZE != 2)
#error "Your I16/U16 doesn't seem to have 16 bits..."
#endif

#if (defined I32SIZE && I32SIZE != 4) || \
    (defined U32SIZE && U32SIZE != 4)
#error "Your I32/U32 doesn't seem to have 32 bits..."
#endif

/*---------------*/
/* some defaults */
/*---------------*/

#ifndef CBC_DEFAULT_PTR_SIZE
#define CBC_DEFAULT_PTR_SIZE    CTLIB_POINTER_SIZE
#else
#if     CBC_DEFAULT_PTR_SIZE != 1 && \
        CBC_DEFAULT_PTR_SIZE != 2 && \
        CBC_DEFAULT_PTR_SIZE != 4 && \
        CBC_DEFAULT_PTR_SIZE != 8
#error "CBC_DEFAULT_PTR_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_ENUM_SIZE
#define CBC_DEFAULT_ENUM_SIZE    sizeof( int )
#else
#if     CBC_DEFAULT_ENUM_SIZE != 0 && \
        CBC_DEFAULT_ENUM_SIZE != 1 && \
        CBC_DEFAULT_ENUM_SIZE != 2 && \
        CBC_DEFAULT_ENUM_SIZE != 4 && \
        CBC_DEFAULT_ENUM_SIZE != 8
#error "CBC_DEFAULT_ENUM_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_INT_SIZE
#define CBC_DEFAULT_INT_SIZE    CTLIB_int_SIZE
#else
#if     CBC_DEFAULT_INT_SIZE != 1 && \
        CBC_DEFAULT_INT_SIZE != 2 && \
        CBC_DEFAULT_INT_SIZE != 4 && \
        CBC_DEFAULT_INT_SIZE != 8
#error "CBC_DEFAULT_INT_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_CHAR_SIZE
#define CBC_DEFAULT_CHAR_SIZE    CTLIB_char_SIZE
#else
#if     CBC_DEFAULT_CHAR_SIZE != 1 && \
        CBC_DEFAULT_CHAR_SIZE != 2 && \
        CBC_DEFAULT_CHAR_SIZE != 4 && \
        CBC_DEFAULT_CHAR_SIZE != 8
#error "CBC_DEFAULT_CHAR_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_SHORT_SIZE
#define CBC_DEFAULT_SHORT_SIZE    CTLIB_short_SIZE
#else
#if     CBC_DEFAULT_SHORT_SIZE != 1 && \
        CBC_DEFAULT_SHORT_SIZE != 2 && \
        CBC_DEFAULT_SHORT_SIZE != 4 && \
        CBC_DEFAULT_SHORT_SIZE != 8
#error "CBC_DEFAULT_SHORT_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_LONG_SIZE
#define CBC_DEFAULT_LONG_SIZE    CTLIB_long_SIZE
#else
#if     CBC_DEFAULT_LONG_SIZE != 1 && \
        CBC_DEFAULT_LONG_SIZE != 2 && \
        CBC_DEFAULT_LONG_SIZE != 4 && \
        CBC_DEFAULT_LONG_SIZE != 8
#error "CBC_DEFAULT_LONG_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_LONG_LONG_SIZE
#define CBC_DEFAULT_LONG_LONG_SIZE    CTLIB_long_long_SIZE
#else
#if     CBC_DEFAULT_LONG_LONG_SIZE != 1 && \
        CBC_DEFAULT_LONG_LONG_SIZE != 2 && \
        CBC_DEFAULT_LONG_LONG_SIZE != 4 && \
        CBC_DEFAULT_LONG_LONG_SIZE != 8
#error "CBC_DEFAULT_LONG_LONG_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_FLOAT_SIZE
#define CBC_DEFAULT_FLOAT_SIZE    CTLIB_float_SIZE
#else
#if     CBC_DEFAULT_FLOAT_SIZE != 1  && \
        CBC_DEFAULT_FLOAT_SIZE != 2  && \
        CBC_DEFAULT_FLOAT_SIZE != 4  && \
        CBC_DEFAULT_FLOAT_SIZE != 8  && \
        CBC_DEFAULT_FLOAT_SIZE != 12 && \
        CBC_DEFAULT_FLOAT_SIZE != 16
#error "CBC_DEFAULT_FLOAT_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_DOUBLE_SIZE
#define CBC_DEFAULT_DOUBLE_SIZE    CTLIB_double_SIZE
#else
#if     CBC_DEFAULT_DOUBLE_SIZE != 1  && \
        CBC_DEFAULT_DOUBLE_SIZE != 2  && \
        CBC_DEFAULT_DOUBLE_SIZE != 4  && \
        CBC_DEFAULT_DOUBLE_SIZE != 8  && \
        CBC_DEFAULT_DOUBLE_SIZE != 12 && \
        CBC_DEFAULT_DOUBLE_SIZE != 16
#error "CBC_DEFAULT_DOUBLE_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_LONG_DOUBLE_SIZE
#define CBC_DEFAULT_LONG_DOUBLE_SIZE    CTLIB_long_double_SIZE
#else
#if     CBC_DEFAULT_LONG_DOUBLE_SIZE != 1  && \
        CBC_DEFAULT_LONG_DOUBLE_SIZE != 2  && \
        CBC_DEFAULT_LONG_DOUBLE_SIZE != 4  && \
        CBC_DEFAULT_LONG_DOUBLE_SIZE != 8  && \
        CBC_DEFAULT_LONG_DOUBLE_SIZE != 12 && \
        CBC_DEFAULT_LONG_DOUBLE_SIZE != 16
#error "CBC_DEFAULT_LONG_DOUBLE_SIZE is invalid!"
#endif
#endif

#ifndef CBC_DEFAULT_ALIGNMENT
#define CBC_DEFAULT_ALIGNMENT    1
#elif   CBC_DEFAULT_ALIGNMENT != 1  && \
        CBC_DEFAULT_ALIGNMENT != 2  && \
        CBC_DEFAULT_ALIGNMENT != 4  && \
        CBC_DEFAULT_ALIGNMENT != 8  && \
        CBC_DEFAULT_ALIGNMENT != 16
#error "CBC_DEFAULT_ALIGNMENT is invalid!"
#endif

#ifndef CBC_DEFAULT_COMPOUND_ALIGNMENT
#define CBC_DEFAULT_COMPOUND_ALIGNMENT    1
#elif   CBC_DEFAULT_COMPOUND_ALIGNMENT != 1  && \
        CBC_DEFAULT_COMPOUND_ALIGNMENT != 2  && \
        CBC_DEFAULT_COMPOUND_ALIGNMENT != 4  && \
        CBC_DEFAULT_COMPOUND_ALIGNMENT != 8  && \
        CBC_DEFAULT_COMPOUND_ALIGNMENT != 16
#error "CBC_DEFAULT_COMPOUND_ALIGNMENT is invalid!"
#endif

#ifndef CBC_DEFAULT_ENUMTYPE
#define CBC_DEFAULT_ENUMTYPE   ET_INTEGER
#endif

#if ARCH_NATIVE_BYTEORDER == ARCH_BYTEORDER_BIG_ENDIAN
#define CBC_NATIVE_BYTEORDER   CBO_BIG_ENDIAN
#elif ARCH_NATIVE_BYTEORDER == ARCH_BYTEORDER_LITTLE_ENDIAN
#define CBC_NATIVE_BYTEORDER   CBO_LITTLE_ENDIAN
#else
#error "unknown native byte order"
#endif

#ifndef CBC_DEFAULT_BYTEORDER
#define CBC_DEFAULT_BYTEORDER  CBC_NATIVE_BYTEORDER
#endif

/*--------------------------------------*/
/* macros for different checks/warnings */
/*--------------------------------------*/

#if defined G_WARN_ON && defined G_WARN_ALL_ON
#define PERL_WARNINGS_ON (PL_dowarn & (G_WARN_ON | G_WARN_ALL_ON))
#else
#define PERL_WARNINGS_ON  PL_dowarn
#endif

#define WARN(args) STMT_START { if (PERL_WARNINGS_ON) Perl_warn args; } STMT_END

#define WARN2(args) STMT_START { if (PERL_WARNINGS_ON && THIS->cfg.issue_warnings) Perl_warn args; } STMT_END

#define WARN_UNSAFE(type) \
          WARN((aTHX_ "Unsafe values used in %s('%s')", method, type))

#define WARN_FLAGS(type, flags)                                                \
          STMT_START {                                                         \
            if ((flags) & T_UNSAFE_VAL)                                        \
              WARN_UNSAFE(type);                                               \
          } STMT_END

#define CROAK_UNDEF_STRUCT(ptr)                                                \
          Perl_croak(aTHX_ "Got no definition for '%s %s'",                    \
                           (ptr)->tflags & T_UNION ? "union" : "struct",       \
                           (ptr)->identifier)

#define WARN_UNDEF_STRUCT(ptr)                                                 \
          WARN((aTHX_ "Got no definition for '%s %s'",                         \
                      (ptr)->tflags & T_UNION ? "union" : "struct",            \
                      (ptr)->identifier))

/*----------------------------*/
/* checks if an SV is defined */
/*----------------------------*/

#define DEFINED(sv) ((sv) != NULL && SvOK(sv))

/*----------------------------------*/
/* avoid warnings with older perl's */
/*----------------------------------*/

#if PERL_REVISION == 5 && PERL_VERSION < 6
# define CONST_CHAR(x) ((char *)(x))
#else
# define CONST_CHAR(x) (x)
#endif

/*--------------------------------------------------*/
/* macros to create SV's/HV's with constant strings */
/*--------------------------------------------------*/

#define NEW_SV_PV_CONST(str) \
          newSVpvn(str, sizeof(str)/sizeof(char)-1)

#define HV_STORE_CONST(hash, key, value)                                       \
        STMT_START {                                                           \
          SV *_val = value;                                                    \
          if (hv_store(hash, key, sizeof(key)/sizeof(char)-1, _val, 0) == NULL)\
            SvREFCNT_dec(_val);                                                \
        } STMT_END

/*-------------------------*/
/* get the size of an enum */
/*-------------------------*/

#define GET_ENUM_SIZE(pCfg, pES)                                               \
          ((pCfg)->layout.enum_size > 0                                        \
            ? (unsigned) (pCfg)->layout.enum_size                              \
            : (pES)->sizes[-(pCfg)->layout.enum_size]) 

/*------------------------------------------------*/
/* this is needed quite often for unnamed structs */
/*------------------------------------------------*/

#define FOLLOW_AND_CHECK_TSPTR(pTS)                                            \
        STMT_START {                                                           \
          if ((pTS)->tflags & T_TYPE)                                          \
          {                                                                    \
            Typedef *_pT = (Typedef *) (pTS)->ptr;                             \
            for(;;)                                                            \
            {                                                                  \
              if (_pT && _pT->pType->tflags & T_TYPE                           \
                      && _pT->pDecl->pointer_flag == 0                         \
                      && _pT->pDecl->array_flag == 0)                          \
                _pT = (Typedef *) _pT->pType->ptr;                             \
              else                                                             \
                break;                                                         \
            }                                                                  \
            (pTS) = _pT->pType;                                                \
          }                                                                    \
                                                                               \
          if (((pTS)->tflags & T_COMPOUND) == 0)                               \
            fatal("Unnamed member was not struct or union (type=0x%08X) "      \
                  "in %s line %d", (pTS)->tflags, __FILE__, __LINE__);         \
                                                                               \
          if ((pTS)->ptr == NULL)                                              \
            fatal("Type pointer to struct/union was NULL in %s line %d",       \
                  __FILE__, __LINE__);                                         \
        } STMT_END


/*===== TYPEDEFS =============================================================*/

typedef struct {

  CParseConfig  cfg;
  CParseInfo    cpi;

  enum {
    ET_INTEGER, ET_STRING, ET_BOTH
  }             enumType;

  /* boolean options */
  unsigned      order_members      : 1;

  const char   *ixhash;
  HV           *hv;
  BasicTypes    basic;

} CBC;


/*===== FUNCTION PROTOTYPES ==================================================*/

#endif
